clap_derive/src/derives/arg_enum.rs

```diff
@@ -7,86 +7,98 @@
+    variants
         .iter()
-        .map(|ref variant| String::from(variant.ident.as_ref()))
-        .collect::<Vec<_>>();
+        .flat_map(|variant| {
+            let attrs = Attrs::from_struct(
+                variant.span(),
+                &variant.attrs,
+                Name::Derived(variant.ident.clone()),
+                parent_attribute.casing(),
+                parent_attribute.env_casing(),
+            );
 
-    // All of these need to be iterators.
-    let ident_slice = [ident.clone()];
-    let idents = ident_slice.iter().cycle();
+            let mut ret = vec![(attrs.cased_name(), variant.ident.clone())];
 
-    let for_error_message = strings.clone();
+            attrs
+                .enum_aliases()
+                .into_iter()
+                .for_each(|x| ret.push((x, variant.ident.clone())));
 
-    let condition_function_slice = [match is_case_sensitive {
-        true => quote! { str::eq },
-        false => quote! { ::std::ascii::AsciiExt::eq_ignore_ascii_case },
-    }];
-    let condition_function = condition_function_slice.iter().cycle();
+            ret
+        })
+        .collect::<Vec<_>>()
+}
 
-    Ok(quote! {
-        impl ::std::str::FromStr for #ident {
-            type Err = String;
+fn gen_variants(lits: &[(TokenStream, Ident)]) -> TokenStream {
+    let lit = lits.iter().map(|l| l.0.clone()).collect::<Vec<_>>();
 
-            fn from_str(input: &str) -> ::std::result::Result<Self, Self::Err> {
-                match input {
-                    #(val if #condition_function(val, #strings) => Ok(#idents::#variants),)*
-                    _ => Err({
-                        let v = #for_error_message;
-                        format!("valid values: {}",
-                            v.join(" ,"))
-                    }),
-                }
-            }
-        }
-    })
+    quote! {
+        const VARIANTS: &'static [&'static str] = &[#(#lit),*];
+    }
 }
 
-fn impl_variants(ast: &syn::DeriveInput) -> proc_macro2::TokenStream {
-    let ident = &ast.ident;
-    let variants = variants(ast)?
-        .iter()
-        .map(|ref variant| String::from(variant.ident.as_ref()))
-        .collect::<Vec<_>>();
-    let length = variants.len();
+fn gen_from_str(lits: &[(TokenStream, Ident)]) -> TokenStream {
+    let (lit, variant): (Vec<TokenStream>, Vec<Ident>) = lits.iter().cloned().unzip();
+
+    quote! {
+        fn from_str(input: &str, case_insensitive: bool) -> ::std::result::Result<Self, String> {
+            let func = if case_insensitive {
+                ::std::ascii::AsciiExt::eq_ignore_ascii_case
+            } else {
+                str::eq
+            };
 
-    Ok(quote! {
-        impl #ident {
-            fn variants() -> [&'static str; #length] {
-                #variants
+            match input {
+                #(val if func(val, #lit) => Ok(Self::#variant),)*
+                _ => Err(String::from("something went wrong parsing the value")),
```

Comment:
This is unreachable. If it came down to `from_str`, the values have already been checked. Let's empathize the fact.
```suggestion
                e => unreachable!("The impossible variant have been spotted: {}", e),
```